[14.0][MIG] sale_invoice_policy

enriquemartin opened the pull request:
Migration + some changes:
- Simplified field management in `res.config.settings`: removed duplicate fields to use the ones added by sale (`default_invoice_policy`).
- Moved `post-init-hook` to `pre-init-hook` in order to keep the default method of the `default_invoice_policy` field.
- Same literals for `invoice_policy` field as in sale
- Updated pot file + Spanish translation

rousseldenis committed:
[10.0][ADD] sale_invoice_policy

rousseldenis committed:
[IMP] Add search method, secures _get_to_invoice_qty,compute_invoice_status computation

rousseldenis committed:
[IMP] sale_invoice_policy: Implement inverse

rousseldenis committed:
[IMP] sale_invoice_policy: Let the availability of choosing no policy on sale order

rousseldenis committed:
[IMP] sale_invoice_policy: Add config for default_policy and if policy is required on Sale Orders

acsonefho committed:
[FIX] Fix perf issue during compute invoice status (of sale.order.line)

oca-travis committed:
[UPD] Update sale_invoice_policy.pot

OCA-git-bot committed:
[ADD] icon.png

AaronHForgeFlow committed:
[MIG]sale_invoice_policy to v12

oca-travis committed:
[UPD] Update sale_invoice_policy.pot

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: sale-workflow-12.0/sale-workflow-12.0-sale_invoice_policy
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-12-0/sale-workflow-12-0-sale_invoice_policy/

georginaf committed:
Added translation using Weblate (Catalan)

georginaf committed:
Translated using Weblate (Catalan)

Currently translated at 100.0% (17 of 17 strings)

Translation: sale-workflow-12.0/sale-workflow-12.0-sale_invoice_policy
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-12-0/sale-workflow-12-0-sale_invoice_policy/ca/

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: sale-workflow-12.0/sale-workflow-12.0-sale_invoice_policy
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-12-0/sale-workflow-12-0-sale_invoice_policy/

pedrocs-exo committed:
Added translation using Weblate (Portuguese)

pedrocs-exo committed:
Translated using Weblate (Portuguese)

Currently translated at 100.0% (17 of 17 strings)

Translation: sale-workflow-12.0/sale-workflow-12.0-sale_invoice_policy
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-12-0/sale-workflow-12-0-sale_invoice_policy/pt/

renatonlima committed:
[FIX] pep8 and pylint

SandieFavre committed:
Translated using Weblate (French)

Currently translated at 70.6% (12 of 17 strings)

Translation: sale-workflow-12.0/sale-workflow-12.0-sale_invoice_policy
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-12-0/sale-workflow-12-0-sale_invoice_policy/fr/

quentinDupont committed:
Translated using Weblate (French)

Currently translated at 100.0% (17 of 17 strings)

Translation: sale-workflow-12.0/sale-workflow-12.0-sale_invoice_policy
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-12-0/sale-workflow-12-0-sale_invoice_policy/fr/

alexeirivera87 committed:
[IMP] sale_invoice_policy: black, isort, prettier

alexeirivera87 committed:
[MIG] sale_invoice_policy: Migration to 13.0

oca-travis committed:
[UPD] Update sale_invoice_policy.pot

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
sale_invoice_policy 13.0.1.0.1

alan196 committed:
Added translation using Weblate (Spanish (Mexico))

alan196 committed:
Translated using Weblate (Spanish (Mexico))

Currently translated at 100.0% (17 of 17 strings)

Translation: sale-workflow-13.0/sale-workflow-13.0-sale_invoice_policy
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-13-0/sale-workflow-13-0-sale_invoice_policy/es_MX/

enriquemartin committed:
[IMP] sale_invoice_policy: black, isort, prettier

enriquemartin committed:
[MIG] sale_invoice_policy: Migration to 14.0

damdam-s commented:
travis error is not related to this module.

However, can you please create 1 commit containing the "normal" migration to v14 and another one in which you do the evolutions/changes please?

Changes will be easier to review w/ this process.
thanks a lot in advance

enriquemartin commented:
That was my initial idea: do a standard migration and then propose changes. But in the process of reviewing the code,  I encountered some issues that did not work without changes. E.g. the default value of `default_invoice_policy` field is not working on a first installation, because of that the `post_init_hook` has to be changed to `pre_init_hook`. 
I will separate the rest of the changes in a separate commit.
Thanks!

enriquemartin committed:
[IMP] sale_invoice_policy: simplified addon

- Removed duplicate fields to use the ones added by sale (default_invoice_policy).
- Same literals for invoice_policy field as in sale
- Updated pot file + Spanish translation

enriquemartin commented:
Done @damdam-s two separate commits to facilitate review!

enriquemartin commented:
any news on this @damdam-s? thanks in advance!

jonerikceberio reviewed (commented):
LGTM

flotho commented on sale_invoice_policy/models/product_template.py:
```diff
@@ -0,0 +1,55 @@
+        """
+        invoice_policy = self.env.context.get("invoice_policy")
+        for tmpl in self:
+            if tmpl.type != "service" and invoice_policy:
```
Could you please precise why you make a difference with service products.
either 'service' and 'product' could be invoiced based on ordered or delivered quantity.

enriquemartin replied:
You are right, we no longer make a distinction.

enriquemartin replied:
I was not right. We need to make a distinction, for example, on orders with storable product and shipping charges: if we mark the billing policy as delivered quantities on the order, shipping charges will never be billed (unless the delivered quantity of this service is forced).
I reverse this change, and leave a comment in the code.

flotho commented on sale_invoice_policy/models/sale_order.py:
```diff
@@ -0,0 +1,33 @@
+        help="Ordered Quantity: Invoice quantities ordered by the customer.\n"
+        "Delivered Quantity: Invoice quantities delivered to the customer.",
+    )
+    invoice_policy_required = fields.Boolean(default=False, readonly=True)
```
sorry if I didn't read correctly but could you precise where this field is used, I can see a view but no python part using this variable.

enriquemartin replied:
Field removed. Indeed it was not used after code simplification.

flotho commented on sale_invoice_policy/models/sale_order_line.py:
```diff
@@ -0,0 +1,69 @@
+    def _get_to_invoice_qty(self):
+        invoice_policies = set(self.mapped("order_id.invoice_policy"))
+        line_by_id = {sol.id: sol for sol in self}
+        done_lines = self.env["sale.order.line"].browse()
```
won't it return all the order.line of the database?

i-vyshnevska replied:
no that will be empty model this is equal to 
`done_lines = self.env["sale.order.line"]`

enriquemartin replied:
Modified to simplify understanding

enriquemartin commented:
@i-vyshnevska any chance to review this? Thanks in advance!

i-vyshnevska commented on sale_invoice_policy/__manifest__.py:
```diff
@@ -0,0 +1,21 @@
+        "views/product_template_view.xml",
+        "views/sale_view.xml",
+    ],
+    "pre_init_hook": "pre_init_hook",
```
should stay in post_init_hook as in pre-state "invoice_policy" field will not de recomputed with code presented in this addon

enriquemartin replied:
I do not understand the problem. The _pre_init_hook()_ just creates in SQL the new _default_invoice_policy_ column and initializes it with the same value as _invoice_policy_.
It does not need any python code at all.

i-vyshnevska replied:
nevermind, you are right

i-vyshnevska commented on sale_invoice_policy/models/sale_order_line.py:
```diff
@@ -0,0 +1,69 @@
+        return True
+
+    @api.depends(
+        "state",
+        "product_uom_qty",
+        "qty_delivered",
+        "qty_to_invoice",
+        "qty_invoiced",
```
```suggestion
```
dependency on these fields injected in parent method so we don't need to have it here

enriquemartin replied:
Removed!

damdam-s reviewed (approved):
thanks

flotho reviewed (approved):
thanks for your explanations

enriquemartin commented:
In order to be compatible with other addons as `sale_automatic_workflow`, the module cannot set the default value for _invoice_policy_ on `sale.order`, so I have deleted the `default_get` method.
An empty value on field _invoice_policy_ on sale order means that the used value will be the one in each product.

rousseldenis commented on sale_invoice_policy/models/sale_order.py:
```diff
@@ -1,43 +1,22 @@
+        help="Ordered Quantity: Invoice quantities ordered by the customer.\n"
+        "Delivered Quantity: Invoice quantities delivered to the customer.",
     )
-    invoice_policy_required = fields.Boolean(
```
This module changes the standard behaviour. It's better to keep it optional.

enriquemartin replied:
I don't understand what you are talking about: are you meaning the deletion of the `invoice_policy_required` field is not "standard"?
As I said the option to make required `invoice_policy` on sale.order makes this module not compatible with others like `sale_automatic_workflow`.

bunhin replied:
It seems the comment from @rousseldenis has made this PR review process not active, while asked question not yet replied

elvise commented:
Hi @enriquemartin and thanks for your porting ! 
Can you rebase ?
Thanks in advance :)

enriquemartin commented:
@dreispt why the needs fixing tag?

dreispt commented:
@enriquemartin That looks like a mistake. I guess this is ready to merge!

dreispt commented:
/ocabot merge nobump
